contract/models/contract.py

```diff
@@ -0,0 +1,716 @@
+                    field.company_dependent,
+                    field.name in self.NO_SYNC,
+                )
+            ):
+                if self.contract_template_id[field_name]:
+                    self[field_name] = self.contract_template_id[field_name]
+
+    @api.onchange("partner_id", "company_id")
+    def _onchange_partner_id(self):
+        partner = (
+            self.partner_id
+            if not self.company_id
+            else self.partner_id.with_company(self.company_id)
+        )
+        self.pricelist_id = partner.property_product_pricelist.id
+        self.fiscal_position_id = partner.env[
+            "account.fiscal.position"
+        ]._get_fiscal_position(partner)
+        if self.contract_type == "purchase":
+            self.payment_term_id = partner.property_supplier_payment_term_id
+        else:
+            self.payment_term_id = partner.property_payment_term_id
+        self.invoice_partner_id = self.partner_id.address_get(["invoice"])["invoice"]
+
+    def _convert_contract_lines(self, contract):
+        self.ensure_one()
+        new_lines = self.env["contract.line"]
+        contract_line_model = self.env["contract.line"]
+        for contract_line in contract.contract_line_ids:
+            vals = contract_line._convert_to_write(contract_line.read()[0])
+            # Remove template link field
+            vals.pop("contract_template_id", False)
+            vals["date_start"] = fields.Date.context_today(contract_line)
+            vals["recurring_next_date"] = fields.Date.context_today(contract_line)
+            new_lines += contract_line_model.new(vals)
+        new_lines._onchange_is_auto_renew()
+        return new_lines
+
+    def _prepare_invoice(self, date_invoice, journal=None):
+        """Prepare in a Form the values for the generated invoice record.
+
+        :return: A tuple with the vals dictionary and the Form with the
+          preloaded values for being used in lines.
+        """
+        self.ensure_one()
+        if not journal:
+            journal = (
+                self.journal_id
+                if self.journal_id.type == self.contract_type
+                else self.env["account.journal"].search(
+                    [
+                        ("type", "=", self.contract_type),
+                        ("company_id", "=", self.company_id.id),
+                    ],
+                    limit=1,
+                )
+            )
+        if not journal:
+            raise ValidationError(
+                _(
+                    "Please define a %(contract_type)s journal "
+                    "for the company '%(company)s'."
+                )
+                % {
+                    "contract_type": self.contract_type,
+                    "company": self.company_id.name or "",
+                }
+            )
+        invoice_type = "out_invoice"
+        if self.contract_type == "purchase":
+            invoice_type = "in_invoice"
+        move_form = Form(
```

Comment:
Not wanting to push a new debate on that, I'm convinced this is not a correct solution

Reply:
I'm not as you know. With this, we assure maximum compatibility. Even with `onchange_helper`, you won't get the same compatibility. I know it's not the most efficient one, but now on v16, the performance gets better.